File: Heaps_Priority-queues/monk_and_some_queries/monk-query.py
import queue

t = int(input())
maxPq = queue.PriorityQueue()
minPq = queue.PriorityQueue()
aux = {}

for i in range(t):
    enter = list(map(int, input().split()))

    if enter[0] == 1:
        # insert element
        num = enter[1]
        maxPq.put(-num)
        minPq.put(num)

        if num not in aux:
            aux[num] = 1
        else:
            aux[num] += 1
    
    elif enter[0] == 2:
        # remove element
        num = enter[1]

        if num not in aux:
            print('-1')
        elif aux[num] == 1:
            aux.pop(num)
        else:
            aux[num] -= 1

    elif enter[0] == 3:
        # max element
        tmp = maxPq.get()
        while(-tmp not in aux and not maxPq.empty()):
            tmp = maxPq.get()
        
        if maxPq.empty():
            print('-1')
        else:
            print(-tmp)
            maxPq.put(tmp)

    
    else:
        # min element
        tmp = minPq.get()
        while(tmp not in aux and not minPq.empty()):
            tmp = minPq.get()
        
        if minPq.empty():
            print('-1')
        else:
            print(tmp)
            minPq.put(tmp)
            

        

File: DataStructure/Heaps_Priority-queues/monk_and_some_queries/monk-query.cpp
#include <bits/stdc++.h>
using namespace std;
#define LL long long int

int main(){
    priority_queue<int> mxPq;
    priority_queue<int> minPq;
    int a[100002];

    // complete the array with zeroes;
    memset(a, 0, sizeof(a));

    int query_size;
    int opt;
    int num;

    cin >> query_size;

    while(query_size--){
        cin >> opt;

        switch(opt){
            case 1:
                // add
                cin >> num;
                mxPq.push(num);
                minPq.push(-num);
                a[num] ++;
                break;

            case 2:
                // remove
                cin >> num;
                if(a[num])
                    a[num] --;
                else
                    cout << -1 << endl;

                break;

            case 3:
                // max
                while(mxPq.size() && a[mxPq.top()] == 0){
                    // this number don't exist any more
                    mxPq.pop();
                }

                if(mxPq.size()){
                    //has element
                    cout << mxPq.top() << endl;
                }else{
                    // empty
                    cout << -1 << endl;
                }
                break;

            case 4:
                // min
                while(minPq.size() && a[-minPq.top()] == 0){
                    // this number don't exist any more
                    minPq.pop();
                }

                if (minPq.size()){
                    cout << -minPq.top() << endl;
                
                }else{
                    cout << -1 << endl;
                }
                break;
        }
    }
    
    return 0;
}